src/functions_math.jl

```diff
@@ -43,8 +43,12 @@ function Base.:*(a::NamedDimsArray{L,T,2,<:CoVector}, b::AbstractVector) where {
     return *(parent(a), b)
 end
 
-# Using `CovVector` results in Method ambiguities; have to define more specific methods.
-for A in (Adjoint{<:Any, <:AbstractVector}, Transpose{<:Real, <:AbstractVector{<:Real}})
+# Using `CoVector` results in Method ambiguities; have to define more specific methods.
+specifics = (Adjoint{<:Any, <:AbstractVector},
+             Adjoint{<:Any, <:AbstractMatrix{T1}} where T1,
+             Transpose{<:Real, <:AbstractVector{<:Real}},
+             Diagonal{<:Any})
+for A in specifics
```

Comment:
Two things are unclear to me:

### Why is this the right way to handle `Diagonal` rather than `@declare_matmul` below ?
In particular, because the others in this loop perform noperation that returns a scalar, and thus doesn't need to worry about preserving names, where as for `Diagonal`  that is not the case.

```julia
julia> [10, 20, 30]' * [1,2,3]
140
```
vs
```julia
julia> Diagonal([1,2]) * [10, 20]
2-element Array{Int64,1}:
 10
 40
```


### why is this the right way to handle the product of the `Adjoint` of a Matrix with a Vector

Which is again is an operation that does does not return scalar and thus needs to preserve names.
```julia
julia> [10 20; 31 32]' * [1,2]
2-element Array{Int64,1}:
 72
 84
```

Reply:
They were indeed not the right way, thanks for pointing it out. I've added tests which ensure that NamedDimsArray is returned from these operations, rather than an array which was the case in the first implementation.